ocaml/xapi/xapi_pool.ml

```diff
@@ -1745,29 +1745,40 @@ let eject_self ~__context ~host =
       )
       my_vms_with_records ;
     (* all control domains resident on me should be destroyed once I leave the
-       		pool, therefore pick them out as follows: if they have a valid resident_on,
-       		the latter should be me; if they don't (e.g. they are halted), they should have
-       		disks on my local storage *)
+       pool, therefore pick them out as follows: if they have a valid
+       resident_on, the latter should be me; if they don't (e.g. they are
+       halted), they should have disks on my local storage *)
     let vm_is_resident_on_host vm_rec host =
       Db.is_valid_ref __context vm_rec.API.vM_resident_on
       && vm_rec.API.vM_resident_on = host
     in
-    let vm_has_disks_on_local_sr_of_host vm_ref host =
-      let is_sr_local x = not (Helpers.is_sr_shared ~__context ~self:x) in
+    let vm_has_disks_on_local_sr_of_host vm_rec host =
+      let is_sr_local x =
+        try not (Helpers.is_sr_shared ~__context ~self:x) with _ -> false
+      in
       let host_has_sr x =
         Helpers.check_sr_exists_for_host ~__context ~self:x ~host <> None
       in
-      Db.VM.get_VBDs ~__context ~self:vm_ref
-      |> List.map (fun x -> Db.VBD.get_VDI ~__context ~self:x)
-      |> List.filter (fun x -> x <> Ref.null)
-      (* filter out null ref VDIs (can happen e.g. for CDs) *)
-      |> List.map (fun x -> Db.VDI.get_SR ~__context ~self:x)
+      let ignore_invalid_ref f x =
+        try
+          let ref = f x in
+          if ref <> Ref.null then
+            Some ref
+          else
+            None
+        with _ -> None
```

Comment:
Same here, perhaps we should have a Helpers.ignore_invalid_ref_exn and a Ref.to_option, the former would ignore exceptions (see above, just the specific one we want to ignore), and the other one would just do the Ref.null -> None conversion.
Then we could perhaps reuse this in other places in the code.